Tech 97 input

HendrikThePendric opened the pull request:
I've introduced a whole bunch of form fields here:
- **CheckBox** and **Switch**: both a checkbox really but with a different appearance
- **RadioButton** and **RadioButtonGroup**: because there is no use for radio-buttons by themselves
- **SelectField**: a dropdown, supporting custom and native dropdown elements
- **TextField**: This has a lot of options and a huge CSS file. It can be used to display inputs of type text/search/password/number/etc. but it's also used by the SelectField

The above are to be part of the public API. But I also introduced the **ToggleField** and **FormField** components, which are only used internally by the others.

And, whilst in this branch, I noticed a LOT of things were broken in non-webkit browsers. So this branch also includes quite a few style fixes. I managed to fix almost all cross-browser issues, but for some IE11 rounded corner/ border-radius issues, I reverted to a _graceful degradation_ strategy. I'll mark the relevant files. If you can think of a solution, please let me know.

HendrikThePendric commented on src/Button/button.css:
```diff
@@ -262,12 +265,43 @@
+    /* Graceful degradation: remove the rounded corners on all buttons apart from circle.
+       These are not rendered correctly in IE11, where they produce white pixels in the corners  */
+    .d2ui-button {
+        border-radius: 0;
```
If you disable this line and open the example app in IE11 you will see some white pixels in the rounded corners. Maybe you know of a way to get rid of them.

HendrikThePendric commented on src/Input/textfield.css:
```diff
@@ -0,0 +1,342 @@
+/* IE11 hacks */
+@media all and (-ms-high-contrast: none) {
+    .d2ui-text-field {
+        border-radius: 0;
```
Again... If you disable this line and open the example app in IE11 you will see some white pixels in the rounded corners. Maybe you know of a way to get rid of them.

varl commented on example-app/src/InputDemo.js:
```diff
@@ -0,0 +1,169 @@
+                variant="outlined"
+                native
+                includeEmpty
+                emptyOptionText="None"
```
```
const selectOptions = [
    { value: null, label: "None" },
    { value: 1, label: 'One' },
    { value: 2, label: 'Two' },
    { value: 3, label: 'Three' },
];
```

HendrikThePendric replied:
You are right that two props for this are complete overkill. However, I think your last suggestion of just including it in the options is not the best way to go. 

First of all because I am making some exceptions for the empty option in the component, so it would be good to keep it separated.

But perhaps more importantly, I think keeping it separate will make for a nicer API. This ways you can use option lists that you receive (from the API) directly. Otherwise, you will continuously have to do stuff like 
```javascript
const options = [ { value: null, label: 'None' }, ...someCollectionAsArray ]
```

HendrikThePendric replied:
Now implemented as one property: `emptyOption` (string)

varl commented:
![image](https://user-images.githubusercontent.com/185449/45494057-8297b280-b770-11e8-9cbf-78367826c542.png)

When writing long help texts they just force the parent to become wider, according to the MD docs the helper text should just wrap below the input:

![image](https://user-images.githubusercontent.com/185449/45494108-aa871600-b770-11e8-8a97-386624edede2.png)

varl commented on src/Input/FormField.js:
```diff
@@ -0,0 +1,44 @@
+    error,
+    dense,
+    fullWidth,
+    helpText,
```
I might be going overboard here, but following the MD they call the different labels:

* Label text
* Assistive text

The first is translated to `label` in property form, so I'm thinking we might do the same with the second: `assistive` (annoying to write) or `assist`. Then I note that @joeDHIS  calls it "Helper text", so maybe just `helper` is sufficient? Or `info`?

I vote for `helper`.

HendrikThePendric replied:
I'd like to keep `helpText` here because `helper` is often used to describe utility functions, so I am afraid users might start passing functions into it. At the same time, I think that `info` is not descriptive enough. It works for a default state, because then that is what it is: some text that explains what to do with the input. However, once the component goes into _error_ or _warning_ state, then actually the text is there to **help** the user correct its input.

I just think that helpText is wonderfully apt for this situation:
- It is **text**
- It **helps** the user fill in the input correctly.

HendrikThePendric replied:
implemented as `helperText`

varl commented on src/Input/FormField.js:
```diff
@@ -0,0 +1,44 @@
+    disabled,
+    error,
+    dense,
+    fullWidth,
```
Does `fullWidth` just mean we turn it into a `block`? Could we just call it `block`? :)

varl commented:
![image](https://user-images.githubusercontent.com/185449/45494987-9d6b2680-b772-11e8-9814-1a268982363a.png)

RTL is mostly OK, but there are some quirks for the text fields.

HendrikThePendric commented:
Below is a summary of how I implemented the PR feedback:
- Updated props: fullWidth => block & includeEmpty + emptyOptionText => emptyOption (kept helpText)
- Implemented dense layout on CheckBox, RadioButton & Switch
- Extracted Field from TextField
- TextField and SelectField now just supply an inputComponent to Field for their styles
- Implemented TextField multiline property, which will render a <textarea />
- Renamed FormField to FieldWrap to avoid confusion with Field component
- Introduced required property and helper function to append asterisks to label texts

varl commented:
Changes approved.
